File: src/services/finpass-recovery.service.ts
/**
 * Financial Password Recovery Service
 *
 * Manages user requests to reset their financial password
 * SLA: 3-5 business days for manual admin processing
 *
 * Process:
 * 1. User creates request → BLOCKS ALL EARNINGS → notifies all admins
 * 2. Admin conducts video verification OUTSIDE bot (Telegram/WhatsApp/etc)
 * 3. Admin reviews → changes status to 'in_review'
 * 4. Admin approves → generates new password, sends to user
 * 5. User receives new password via bot (1-hour TTL for repeat view)
 * 6. User successfully uses new password → UNBLOCKS EARNINGS
 *
 * CRITICAL Security:
 * - ALL earnings blocked from request creation until first successful password use
 * - Prevents unauthorized withdrawals during recovery period
 * - User must prove possession of new password before earnings resume
 *
 * Anti-abuse:
 * - Unique constraint: only one open request per user
 * - Optional: Add cooldown period (e.g., 30 days between requests)
 *
 * Video Verification:
 * - Conducted OUTSIDE bot via external channels (Telegram DM, WhatsApp, etc)
 * - Fields video_required/video_verified are admin markers only
 * - Bot does not handle video upload/storage
 */

import { AppDataSource } from '../database/data-source';
import { FinancialPasswordRecovery, User } from '../database/entities';
import { createLogger, logAdminAction } from '../utils/logger.util';
import { generateFinancialPassword, hashPassword } from '../utils/crypto.util';
import { notificationService } from './notification.service';
import Redis from 'ioredis';
import { config } from '../config';

const logger = createLogger('FinpassRecoveryService');

// Redis client for 1-hour password reveal TTL (same pattern as registration)
const redis = new Redis({
  host: config.redis.host,
  port: config.redis.port,
  password: config.redis.password,
  db: config.redis.db,
});

export class FinpassRecoveryService {
  private repo = AppDataSource.getRepository(FinancialPasswordRecovery);
  private userRepo = AppDataSource.getRepository(User);

  /**
   * Create recovery request from user
   * Prevents duplicate open requests via unique constraint
   *
   * CRITICAL: Blocks ALL earnings for user during recovery period
   * Earnings will be unblocked only after user successfully uses new password
   *
   * @returns requestId if successful
   */
  async createRequest(userId: number): Promise<{ success: boolean; error?: string; requestId?: number }> {
    try {
      // Check for existing open request
      const existing = await this.repo.findOne({
        where: [
          { user_id: userId, status: 'pending' },
          { user_id: userId, status: 'in_review' },
          { user_id: userId, status: 'approved' },
        ],
      });

      if (existing) {
        // Return existing request ID (idempotent)
        logger.info('User already has open finpass recovery request', {
          userId,
          requestId: existing.id,
          status: existing.status,
        });
        return { success: true, requestId: existing.id };
      }

      // CRITICAL: Block earnings during recovery period
      const user = await this.userRepo.findOne({ where: { id: userId } });
      if (!user) {
        return { success: false, error: 'Пользователь не найден' };
      }

      user.earnings_blocked = true;
      await this.userRepo.save(user);

      logger.warn('Earnings blocked for user during finpass recovery', {
        userId,
        telegram_id: user.telegram_id,
      });

      // Create new request
      const request = this.repo.create({
        user_id: userId,
        status: 'pending',
        video_required: true, // Note: video verification conducted outside bot
        video_verified: false,
      });

      const saved = await this.repo.save(request);

      // Notify all admins
      await notificationService.notifyAllAdmins(
        'Заявка на восстановление финпароля',
        `👤 Пользователь ID: ${userId}\n` +
        `🆔 Заявка: #${saved.id}\n\n` +
        `⏳ SLA: 3–5 рабочих дней\n` +
        `Требуется ручная проверка и сброс пароля.`
      ).catch(err => {
        logger.error('Failed to notify admins about finpass recovery request', {
          requestId: saved.id,
          error: err,
        });
      });

      logger.info('Finpass recovery request created', {
        userId,
        requestId: saved.id,
      });

      return { success: true, requestId: saved.id };
    } catch (error) {
      logger.error('Error creating finpass recovery request', {
        userId,
        error: error instanceof Error ? error.message : String(error),
      });
      return {
        success: false,
        error: 'Не удалось создать заявку. Попробуйте позже.',
      };
    }
  }

  /**
   * List pending recovery requests (for admin panel)
   *
   * @param limit Max number of requests to return
   */
  async listPending(limit = 20): Promise<FinancialPasswordRecovery[]> {
    return this.repo.find({
      where: [
        { status: 'pending' },
        { status: 'in_review' },
        { status: 'approved' },
      ],
      relations: ['user'],
      order: { created_at: 'ASC' }, // FIFO: oldest first
      take: limit,
    });
  }

  /**
   * Get single request by ID
   */
  async getRequest(requestId: number): Promise<FinancialPasswordRecovery | null> {
    return this.repo.findOne({
      where: { id: requestId },
      relations: ['user', 'processed_by_admin'],
    });
  }

  /**
   * Admin takes request for review
   */
  async takeInReview(requestId: number, adminId: number): Promise<boolean> {
    try {
      const request = await this.repo.findOne({ where: { id: requestId } });
      if (!request) return false;
      if (request.status === 'rejected' || request.status === 'sent') return false;

      request.status = 'in_review';
      request.processed_by_admin_id = adminId;
      await this.repo.save(request);

      logAdminAction(adminId, 'finpass_in_review', { requestId });
      logger.info('Finpass recovery request in review', {
        requestId,
        adminId,
      });

      return true;
    } catch (error) {
      logger.error('Error taking finpass request in review', {
        requestId,
        adminId,
        error,
      });
      return false;
    }
  }

  /**
   * Admin rejects request
   */
  async reject(requestId: number, adminId: number, comment?: string): Promise<boolean> {
    try {
      const request = await this.repo.findOne({ where: { id: requestId } });
      if (!request) return false;
      if (request.status === 'sent') return false; // Can't reject if already sent

      request.status = 'rejected';
      request.processed_by_admin_id = adminId;
      request.processed_at = new Date();
      request.admin_comment = comment || undefined;
      await this.repo.save(request);

      logAdminAction(adminId, 'finpass_reject', { requestId, comment });
      logger.info('Finpass recovery request rejected', {
        requestId,
        adminId,
        comment,
      });

      return true;
    } catch (error) {
      logger.error('Error rejecting finpass request', {
        requestId,
        adminId,
        error,
      });
      return false;
    }
  }

  /**
   * Admin approves and resets financial password
   *
   * Process:
   * 1. Generate new password (6-digit PIN)
   * 2. Hash with bcrypt
   * 3. Update user record
   * 4. Store plain password in Redis (1-hour TTL for repeat view)
   * 5. Send password to user via bot
   * 6. Mark request as 'sent'
   */
  async approveAndReset(requestId: number, adminId: number): Promise<{ success: boolean; error?: string }> {
    try {
      const request = await this.repo.findOne({
        where: { id: requestId },
        relations: ['user'],
      });

      if (!request || !request.user) {
        return { success: false, error: 'Заявка не найдена' };
      }

      if (request.status === 'sent') {
        return { success: false, error: 'Пароль уже был отправлен' };
      }

      if (request.status === 'rejected') {
        return { success: false, error: 'Заявка отклонена' };
      }

      // Get user
      const user = await this.userRepo.findOne({ where: { id: request.user_id } });
      if (!user) {
        return { success: false, error: 'Пользователь не найден' };
      }

      // Generate new password
      const plainPassword = generateFinancialPassword();
      const hashedPassword = await hashPassword(plainPassword);

      // Update user
      user.financial_password = hashedPassword;
      await this.userRepo.save(user);

      // Store plain password in Redis for 1-hour repeat view (same pattern as registration)
      const redisKey = `password:plain:${user.id}`;
      await redis.setex(redisKey, 3600, plainPassword); // 1 hour TTL

      // Update request
      request.status = 'sent';
      request.processed_by_admin_id = adminId;
      request.processed_at = new Date();
      await this.repo.save(request);

      // Send password to user
      const message = [
        '🔐 **Ваш финансовый пароль сброшен**',
        '',
        'Пожалуйста, сохраните новый пароль. Он потребуется для подтверждения финансовых операций.',
        '',
        `**Новый пароль:** \`${plainPassword}\``,
        '',
        '_⚠️ Пароль доступен повторно в течение **1 часа** через кнопку «Показать пароль ещё раз» в профиле_',
      ].join('\n');

      await notificationService.sendCustomMessage(user.telegram_id, message, {
        parse_mode: 'Markdown',
      }).catch((err: Error) => {
        logger.error('Failed to send finpass reset message to user', {
          userId: user.id,
          requestId,
          error: err,
        });
      });

      logAdminAction(adminId, 'finpass_reset_sent', {
        requestId,
        userId: user.id,
      });

      logger.info('Finpass reset and delivered', {
        requestId,
        userId: user.id,
        adminId,
      });

      return { success: true };
    } catch (error) {
      logger.error('Error approving and resetting finpass', {
        requestId,
        adminId,
        error: error instanceof Error ? error.message : String(error),
      });
      return {
        success: false,
        error: 'Не удалось сбросить пароль. Попробуйте позже.',
      };
    }
  }

  /**
   * Get user's last recovery request (for anti-abuse checks)
   */
  async getLastRequest(userId: number): Promise<FinancialPasswordRecovery | null> {
    return this.repo.findOne({
      where: { user_id: userId },
      order: { created_at: 'DESC' },
    });
  }
}

export const finpassRecoveryService = new FinpassRecoveryService();
export default finpassRecoveryService;

File: app/services/finpass_recovery_service.py
# MCP-MARKER:CREATE:FINPASS_RECOVERY_SERVICE
# MCP-ANCHOR: finpass-recovery-service
# MCP-DEPS: [sqlalchemy, loguru]
# MCP-PROVIDES: FinpassRecoveryService, FinancialRecoveryStatus
# MCP-SUMMARY: Service layer for handling financial password recovery workflow.
"""Finpass recovery service.

Business logic for managing financial password recovery requests.
"""

from __future__ import annotations

from collections.abc import Iterable
from datetime import UTC, datetime
from enum import StrEnum

from loguru import logger
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.financial_password_recovery import (
    FinancialPasswordRecovery,
)
from app.repositories.financial_password_recovery_repository import (
    FinancialPasswordRecoveryRepository,
)
from app.repositories.user_repository import UserRepository


class FinancialRecoveryStatus(StrEnum):
    """Lifecycle states for a financial password recovery request."""

    PENDING = "pending"
    IN_REVIEW = "in_review"
    APPROVED = "approved"
    REJECTED = "rejected"
    SENT = "sent"


# Active statuses that block new requests (NOT including SENT - that's completed)
ACTIVE_USER_STATUSES: tuple[FinancialRecoveryStatus, ...] = (
    FinancialRecoveryStatus.PENDING,
    FinancialRecoveryStatus.IN_REVIEW,
    FinancialRecoveryStatus.APPROVED,
)


class FinpassRecoveryService:
    """Service wrapper around financial password recovery workflow."""

    def __init__(self, session: AsyncSession) -> None:
        """Initialize the service with required repositories."""
        self.session = session
        self.repository = FinancialPasswordRecoveryRepository(session)
        self.user_repository = UserRepository(session)

    async def get_pending_by_user(
        self, user_id: int
    ) -> FinancialPasswordRecovery | None:
        """Return the pending request for a user, if any."""
        requests = await self.repository.get_by_user(
            user_id=user_id,
            status=FinancialRecoveryStatus.PENDING.value,
        )
        return requests[0] if requests else None

    async def has_active_recovery(self, user_id: int) -> bool:
        """Check whether a user already has an active recovery process."""
        stmt = (
            select(FinancialPasswordRecovery.id)
            .where(FinancialPasswordRecovery.user_id == user_id)
            .where(
                FinancialPasswordRecovery.status.in_(
                    [status.value for status in ACTIVE_USER_STATUSES]
                )
            )
            .limit(1)
        )
        result = await self.session.execute(stmt)
        return result.scalar_one_or_none() is not None

    async def create_recovery_request(
        self,
        *,
        user_id: int,
        reason: str,
        video_required: bool = True,
    ) -> FinancialPasswordRecovery:
        """Create a new financial password recovery request."""
        normalized_reason = reason.strip()
        if len(normalized_reason) < 10:
            raise ValueError(
                "Recovery reason must contain at least 10 characters"
            )

        user = await self.user_repository.get_by_id(user_id)
        if not user:
            raise ValueError("User not found")

        if await self.get_pending_by_user(user_id):
            raise ValueError(
                "A pending recovery request already exists for this user"
            )

        if await self.has_active_recovery(user_id):
            raise ValueError(
                "You already have an active recovery request awaiting "
                "completion"
            )

        request = await self.repository.create(
            user_id=user_id,
            reason=normalized_reason,
            status=FinancialRecoveryStatus.PENDING.value,
            video_required=video_required,
            video_verified=not video_required,
        )

        # Block earnings immediately upon creating recovery request
        # This prevents any new earnings while recovery is in progress
        from app.services.user_service import UserService

        user_service = UserService(self.session)
        await user_service.block_earnings(user_id, block=True)
        logger.info(
            "Earnings blocked for user due to finpass recovery request",
            extra={"user_id": user_id, "request_id": request.id},
        )

        logger.info(
            "Created financial password recovery request",
            extra={
                "user_id": user_id,
                "request_id": request.id,
                "video_required": video_required,
            },
        )

        return request

    async def get_all_pending(
        self, limit: int | None = None
    ) -> list[FinancialPasswordRecovery]:
        """Fetch pending requests, optionally limited for admin UI."""
        if limit is None:
            return await self.repository.get_pending_requests()

        stmt = (
            select(FinancialPasswordRecovery)
            .where(
                FinancialPasswordRecovery.status
                == FinancialRecoveryStatus.PENDING.value
            )
            .order_by(FinancialPasswordRecovery.created_at.asc())
            .limit(limit)
        )
        result = await self.session.execute(stmt)
        return list(result.scalars().all())

    async def approve_request(
        self,
        *,
        request_id: int,
        admin_id: int,
        admin_notes: str | None = None,
    ) -> FinancialPasswordRecovery:
        """Approve a recovery request."""
        request = await self._get_request(request_id)
        self._ensure_status(
            request,
            allowed={
                FinancialRecoveryStatus.PENDING,
                FinancialRecoveryStatus.IN_REVIEW,
            },
            action="approve",
        )

        updated = await self.repository.update(
            request.id,
            status=FinancialRecoveryStatus.APPROVED.value,
            processed_by_admin_id=admin_id,
            processed_at=datetime.now(UTC),
            admin_comment=admin_notes,
        )

        logger.info(
            "Approved financial password recovery request",
            extra={
                "request_id": request_id,
                "admin_id": admin_id,
            },
        )
        return updated or request

    async def reject_request(
        self,
        *,
        request_id: int,
        admin_id: int | None = None,
        admin_notes: str | None = None,
    ) -> FinancialPasswordRecovery:
        """
        Reject a recovery request.

        Args:
            request_id: Recovery request ID
            admin_id: Admin ID (optional, None for system rejection)
            admin_notes: Admin notes (optional)
        """
        request = await self._get_request(request_id)
        self._ensure_status(
            request,
            allowed={
                FinancialRecoveryStatus.PENDING,
                FinancialRecoveryStatus.IN_REVIEW,
                FinancialRecoveryStatus.APPROVED,
            },
            action="reject",
        )

        updated = await self.repository.update(
            request.id,
            status=FinancialRecoveryStatus.REJECTED.value,
            processed_by_admin_id=admin_id,
            processed_at=datetime.now(UTC),
            admin_comment=admin_notes,
        )

        logger.info(
            "Rejected financial password recovery request",
            extra={
                "request_id": request_id,
                "admin_id": admin_id,
            },
        )
        return updated or request

    async def mark_sent(
        self,
        *,
        request_id: int,
        admin_id: int,
        admin_notes: str | None = None,
    ) -> FinancialPasswordRecovery:
        """Mark an approved request as sent to the user."""
        request = await self._get_request(request_id)
        self._ensure_status(
            request,
            allowed={FinancialRecoveryStatus.APPROVED},
            action="mark as sent",
        )

        updated = await self.repository.update(
            request.id,
            status=FinancialRecoveryStatus.SENT.value,
            processed_by_admin_id=admin_id,
            processed_at=datetime.now(UTC),
            admin_comment=admin_notes,
        )

        logger.info(
            "Marked financial password recovery request as sent",
            extra={
                "request_id": request_id,
                "admin_id": admin_id,
            },
        )
        return updated or request

    async def get_request_by_id(
        self, request_id: int
    ) -> FinancialPasswordRecovery | None:
        """Get a request by ID."""
        return await self.repository.get_by_id(request_id)

    async def _get_request(
        self, request_id: int
    ) -> FinancialPasswordRecovery:
        request = await self.repository.get_by_id(request_id)
        if not request:
            raise ValueError(f"Recovery request {request_id} not found")
        return request

    @staticmethod
    def _ensure_status(
        request: FinancialPasswordRecovery,
        *,
        allowed: Iterable[FinancialRecoveryStatus],
        action: str,
    ) -> None:
        allowed_values = {status.value for status in allowed}
        if request.status not in allowed_values:
            raise ValueError(

                    f"Cannot {action} request {request.id} "
                    f"in status {request.status}"

            )
